consensus/core/src/hashing/sighash.rs

```diff
@@ -0,0 +1,255 @@
+    if hash_type.is_sighash_anyone_can_pay() {
+        return 0.into();
+    }
+
+    if let Some(sig_op_counts_hash) = reused_values.sig_op_counts_hash {
+        sig_op_counts_hash
+    } else {
+        let mut hasher = TransactionSigningHash::new();
+        for input in tx.tx.inputs.iter() {
+            hasher.write_u8(input.sig_op_count);
+        }
+        let sig_op_counts_hash = hasher.finalize();
+        reused_values.sig_op_counts_hash = Some(sig_op_counts_hash);
+        sig_op_counts_hash
+    }
+}
+
+fn payload_hash(tx: &PopulatedTransaction) -> Hash {
+    if tx.tx.subnetwork_id == SUBNETWORK_ID_NATIVE {
+        return 0.into();
+    }
+
+    // TODO: Right now this branch will never be executed, since payload is disabled
+    // for all non coinbase transactions. Once payload is enabled, the payload hash
+    // should be cached to make it cost O(1) instead of O(tx.inputs.len()).
+    let mut hasher = TransactionSigningHash::new();
+    hasher.write_var_bytes(&tx.tx.payload);
+    hasher.finalize()
+}
+
+fn outputs_hash(
+    tx: &PopulatedTransaction,
+    hash_type: SigHashType,
+    reused_values: &mut SigHashReusedValues,
+    input_index: usize,
+) -> Hash {
+    if hash_type.is_sighash_none() {
+        return 0.into();
+    }
+
+    if hash_type.is_sighash_single() {
+        // If the relevant output exists - return its hash, otherwise return zero-hash
+        if input_index >= tx.outputs().len() {
+            return 0.into();
+        }
+
+        let mut hasher = TransactionSigningHash::new();
+        hash_output(&mut hasher, &tx.outputs()[input_index]);
+        return hasher.finalize();
+    }
+
+    // Otherwise, return hash of all outputs. Re-use hash if available.
+    if let Some(outputs_hash) = reused_values.outputs_hash {
+        outputs_hash
+    } else {
+        let mut hasher = TransactionSigningHash::new();
+        for output in tx.tx.outputs.iter() {
+            hash_output(&mut hasher, output);
+        }
+        let outputs_hash = hasher.finalize();
+        reused_values.outputs_hash = Some(outputs_hash);
+        outputs_hash
+    }
+}
+
+fn hash_outpoint(hasher: &mut impl Hasher, outpoint: TransactionOutpoint) {
+    hasher.update(outpoint.transaction_id);
+    hasher.write_u32(outpoint.index);
+}
+
+fn hash_output(hasher: &mut impl Hasher, output: &TransactionOutput) {
+    hasher.write_u64(output.value);
+    hash_script_public_key(hasher, &output.script_public_key);
+}
+
+fn hash_script_public_key(hasher: &mut impl Hasher, script_public_key: &ScriptPublicKey) {
+    hasher.write_u16(script_public_key.version());
+    hasher.write_var_bytes(script_public_key.script());
+}
+
+pub fn calc_schnorr_signature_hash(
+    tx: &PopulatedTransaction,
+    input_index: usize,
+    hash_type: SigHashType,
+    reused_values: &mut SigHashReusedValues,
+) -> Hash {
+    let input = tx.populated_inputs().nth(input_index).unwrap();
```

Comment:
Perhaps pass the (input, entry) tuple from the caller as well, since `nth` re-evaluates the iterator linearly until reaching `n`

Reply:
I think it's pretty cheap and doesn't justify complicating the API (you'll still need to pass the input_index because it's used in `outputs_hash`)

Reply:
So can we add the following method to `PopulatedTransaction` and use it here:

```
    pub fn populated_input(&self, input_index: usize) -> (&TransactionInput, &UtxoEntry) {
        (&self.tx.inputs[input_index], &self.entries[input_index])
    }
```